docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,354 @@
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole).
+4. New feature flag promoted and eventually removed.
+5. (Outside the scope of this proposal) Use of HVPA removed for other workloads and the preexisting HVPA flag removed.
+
+### Vision Beyond This Enhancement
+This proposal is part of a mid-term vision which enables the operator to select an alternative ShootKapi scaling mode
+for any shoot. Said mode prioritises reliability, and is known to work well even under extreme API request profiles.
+
+The following approach is envisioned (discussion is limited to scaling ShootKapi):
+- The default scaling mode is the one proposed by this GEP.
+- For any shoot, the operator is enabled to select an alternative scaling mode. In this mode, the operator specifies a fixed
+  vertical scale, and a horizontal scale in the form of (MinimumReplicaCount, MaximumReplicaCount). HPA is used
+  to apply the horizontal scale.
```

Comment:
>Is this the current HVPA approach?

No. We do have a fallback at present, but it's a different one. HVPA is behind a feature gate, and when disabled, another scaling mechanism is deployed. For kube-apiserver, I think the fallback was HPA-only, but I'm not sure. I haven't seen it in action, so all I have is a vague recollection of what I've seen in source code.

>there is a risk that we will always have (...) clusters we scale manually

This would not be a regression from the current state. HVPA already necessitates occasional manual exceptions.

Reply:
I removed the "Vision Beyond This Enhancement" section from the document.